Tools/LyTestTools/ly_test_tools/o3de/multi_test_framework.py

```diff
@@ -0,0 +1,766 @@
+    """
+    Abstract base Test class
+    """
+    # Maximum time for run, in seconds.
+    timeout = 180
+    # Test file that this test will run.
+    test_module = None
+    # Attach debugger when running the test, useful for debugging crashes. This should never be True on production.
+    # It's also recommended to switch to SingleTest for debugging in isolation.
+    attach_debugger = False
+    # Wait until a debugger is attached at the startup of the test, this is another way of debugging.
+    wait_for_debugger = False
+
+
+class SingleTest(AbstractTestBase):
+    """
+    Test that will be run alone in one instance, with no parallel or serially batched instances.
+    This should only be used if SharedTest, ParallelTest, or BatchedTest cannot be utilized for the test steps.
+    It may also be used for debugging test issues in isolation.
+    """
+    def __init__(self):
+        # Extra cmdline arguments to supply to the instance for the test.
+        self.extra_cmdline_args = []
+        # Whether to use null renderer, this will override use_null_renderer for the BaseTestSuite if not None.
+        self.use_null_renderer = None
+
+    def setup(self, **args):
+        """
+        User-overrideable setup function, which will run before the test
+        """
+        pass
+
+    def wrap_run(self, **args):
+        """
+        User-overrideable wrapper function, which will run before and after test.
+        Any code before the 'yield' statement will run before the test. With code after yield run after the test.
+        """
+        yield
+
+    def teardown(self, **args):
+        """
+        User-overrideable teardown function, which will run after the test
+        """
+        pass
+
+
+class SharedTest(AbstractTestBase):
+    """
+    Runs both of the following test types:
+    1. Tests that will be run in parallel with tests in multiple instances.
+    2. Tests that will be run as serially batched with other tests in a single instance.
+
+    Minimizes total test run duration by reducing repeated overhead from starting the instance.
+    Does not support per test setup/teardown to avoid creating race conditions.
+    """
+    # Specifies if the test can be batched in the same instance.
+    is_batchable = True
+    # Specifies if the test can be run in multiple instances in parallel.
+    is_parallelizable = True
+
+
+class ParallelTest(SharedTest):
+    """Test that will be run in parallel with tests in multiple instances."""
+    is_batchable = False
+    is_parallelizable = True
+
+
+class BatchedTest(SharedTest):
+    """Test that will be run as serially batched with other tests in a single instance."""
+    is_batchable = True
+    is_parallelizable = False
+
+
+class TestResultException(Exception):
+    """Indicates that an unknown result was found during the tests"""
+    pass
+
+
+class Result(object):
+    """Holds test results for a given program/application."""
+    log_attribute = ''
```

Comment:
This should be declared as instance storage (declared inside `__init__`) and not class storage, or else this can break during parallel execution when multiple instances modify the same attr's data. Perhaps move into an init for ResultType? Appears that it is data storage not intended for modification outside of this class, so it should have a leading underscore.

If this attr isn't defined in an init, it should be documented here.

Reply:
My bad - Since this is storing a primitive type and not an object, this isn't a problem.